ports/esp32/usb.c

```diff
@@ -58,6 +59,7 @@ static void usb_callback_rx(int itf, cdcacm_event_t *event) {
                 ringbuf_put(&stdin_ringbuf, usb_rx_buf[i]);
             }
         }
+        mp_hal_wake_main_task_from_isr();
```

Comment:
Is this USB callback actually an ISR?  Or should we instead be using `vTaskNotifyGive(mp_main_task_handle)` instead?

Reply:
Good question! I've tried to trace it through the code path in micropython/ports/esp32/managed_components/espressif__tinyusb/ - this function is called indictly by TinyUSB's "standard" `tud_cdc_rx_cb` callback which eventually traces back up to the top level TinyUSB task tud_task_ext() / tud_task() - so, no, it's an IRQ it's running directly from a separate FreeRTOS Task.

Reply:
Just pushed the change, `xTaskNotifyGive(mp_main_task_handle);` does provide the same speed up thanks!

Reply:
> which eventually traces back up to the top level TinyUSB task tud_task_ext() / tud_task() - so, no, it's an IRQ it's running directly from a separate FreeRTOS Task.

So is it an IRQ, or is it running at thread-level on a separate task?  I'm guessing the latter.

Reply:
thread-level on a separate task.
`ports/esp32/managed_components/espressif__esp_tinyusb_mod/tusb_tasks.c`
``` C
/**
 * @brief This top level thread processes all usb events and invokes callbacks
 */
static void tusb_device_task(void *arg)
{
    ESP_LOGD(TAG, "tinyusb task started");
    while (1) { // RTOS forever loop
        tud_task();
    }
}

esp_err_t tusb_run_task(void)
{
    // This function is not garanteed to be thread safe, if invoked multiple times without calling `tusb_stop_task`, will cause memory leak
    // doing a sanity check anyway
    ESP_RETURN_ON_FALSE(!s_tusb_tskh, ESP_ERR_INVALID_STATE, TAG, "TinyUSB main task already started");
    // Create a task for tinyusb device stack:
    xTaskCreate(tusb_device_task, "TinyUSB", CONFIG_TINYUSB_TASK_STACK_SIZE, NULL, CONFIG_TINYUSB_TASK_PRIORITY, &s_tusb_tskh);
    ESP_RETURN_ON_FALSE(s_tusb_tskh, ESP_FAIL, TAG, "create TinyUSB main task failed");
    return ESP_OK;
}
```

Reply:
OK, thanks for clarifying.

(I'm surprised it's a simple while-1 loop... hopefully `tud_task()` is a blocking call that waits for events on a FreeRTOS queue.)